crates/ty_python_semantic/src/types/narrow.rs

```diff
@@ -633,6 +633,96 @@ impl<'db, 'ast> NarrowingConstraintsBuilder<'db, 'ast> {
             } else {
                 None
             }
+        } else if lhs_ty.is_union_with_single_valued(self.db) {
+            // If the LHS is a union that contains at least one single-valued arm, we narrow the
+            // single-valued portion from the RHS values, and keep non-single-valued arms unchanged.
+            let rhs_values: Option<Type> = if let Type::StringLiteral(string_literal) = rhs_ty {
+                Some(UnionType::from_elements(
+                    self.db,
+                    string_literal
+                        .iter_each_char(self.db)
+                        .map(Type::StringLiteral),
+                ))
+            } else if let Some(tuple_spec) = rhs_ty.tuple_instance_spec(self.db) {
+                Some(UnionType::from_elements(self.db, tuple_spec.all_elements()))
+            } else {
+                None
+            };
```

Comment:
It looks like we now have three copies of this identical logic -- can we extract a helper function instead?

Actually I wonder if we even need this logic anymore at all, now that our iteration support has improved (and already special cases string literals). It seems like this could all be replaced with something like `rhs_ty.try_iterate(db).ok()?.homogeneous_element_type(db)`?